### PR TITLE
PS-5076: Redo log scrubbing ported to 8.0

### PR DESCRIPTION
**Pull request by dutow**

### CODE REVIEW

File: storage/innobase/handler/ha_innodb.cc [hunk 1]
```diff
@@ -21062,6 +21066,19 @@ static MYSQL_SYSVAR_BOOL(immediate_scrub_data_uncompressed,
     srv_immediate_scrub_data_uncompressed, 0,
     "Enable scrubbing of data", NULL, NULL, FALSE);
 
+static MYSQL_SYSVAR_BOOL(scrub_log, srv_scrub_log,
+     PLUGIN_VAR_OPCMDARG | PLUGIN_VAR_READONLY,
+     "Enable background redo log (ib_logfile0, ib_logfile1...) scrubbing",
+     0, 0, 0);
```
Comment (laurynas-biveinis) [inline storage/innobase/handler/ha_innodb.cc:21072]: some of those 0 must be nullptr and some false

File: storage/innobase/handler/ha_innodb.cc [hunk 2]
```diff
@@ -21062,6 +21066,19 @@ static MYSQL_SYSVAR_BOOL(immediate_scrub_data_uncompressed,
...
+
+static MYSQL_SYSVAR_ULONGLONG(scrub_log_speed, innodb_scrub_log_speed,
+     PLUGIN_VAR_OPCMDARG,
+     "Background redo log scrubbing speed in bytes/sec",
+     NULL, NULL,
```
Comment (laurynas-biveinis) [inline storage/innobase/handler/ha_innodb.cc:21077]:
```suggestion
     nullptr, nullptr,
```

File: storage/innobase/log/log0log.cc [hunk 3]
```diff
@@ -61,6 +61,10 @@ this program; if not, write to the Free Software Foundation, Inc.,
 /** Pointer to the log checksum calculation function. */
 log_checksum_func_t log_checksum_algorithm_ptr;
 
+/* Next log block number to do dummy record filling if no log records written
+ * for a while */
+static ulint next_lbn_to_pad = 0;
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc:66]: Is "lbn" a term used in 8.0 InnoDB?

File: storage/innobase/log/log0log.cc [hunk 4]
```diff
@@ -1153,6 +1163,55 @@ void log_position_collect_lsn_info(const log_t &log, lsn_t *current_lsn,
   ut_a(*current_lsn >= *checkpoint_lsn);
 }
 
-  /* @} */
+static void log_pad_current_log_block(void)
+/*===========================*/
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc:1167]: Please remove

File: storage/innobase/log/log0log.cc [hunk 5]
```diff
@@ -1153,6 +1163,55 @@ void log_position_collect_lsn_info(const log_t &log, lsn_t *current_lsn,
...
+{
+  byte b = MLOG_DUMMY_RECORD;
+  ulint pad_length;
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc:1170]: Declare at first use with const or const auto as applicable

File: storage/innobase/log/log0log.cc [hunk 6]
```diff
@@ -1153,6 +1163,55 @@ void log_position_collect_lsn_info(const log_t &log, lsn_t *current_lsn,
...
+  auto handle = log_buffer_reserve(*log_sys, pad_length);
+  for (i = 0; i < pad_length; i++) {
+    log_buffer_write(*log_sys, handle, &b, 1, log_sys->current_file_lsn);
+  }
+  lsn = log_sys->current_file_lsn;
+  ut_a(lsn % OS_FILE_LOG_BLOCK_SIZE == LOG_BLOCK_HDR_SIZE);
+}
+
+static void log_scrub()
+/*=========*/
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc:1192]: Remove

File: storage/innobase/log/log0log.cc [hunk 7]
```diff
@@ -1153,6 +1163,55 @@ void log_position_collect_lsn_info(const log_t &log, lsn_t *current_lsn,
...
+{
+  log_writer_mutex_enter(*log_sys);
+  ulint cur_lbn = log_block_convert_lsn_to_no(log_sys->current_file_lsn);
```
Comment (laurynas-biveinis) [inline storage/innobase/log/log0log.cc:1195]: const local vars which don't change

File: storage/innobase/srv/srv0start.cc [hunk 8]
```diff
@@ -3408,3 +3423,13 @@ void srv_fatal_error() {
 
   exit(3);
 }
+
+/* @} */
```
Comment (laurynas-biveinis) [inline storage/innobase/srv/srv0start.cc:3427]: Is there an opening {@ ?

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]
OK to merge and address comments later

**Comment by dutow** [comment]
https://ps.cd.percona.com/job/percona-server-8.0-param/38

**Pull request merged**